File: Fourth Semester - Design and Programming Environments/C#/ProiectC#/repository/impl/TripRepositoryImpl.cs
using System.Data;
using log4net;
using TravelAgency.exception;
using TravelAgency.model;
using TravelAgency.model.validator;

namespace TravelAgency.repository.database;

public class TripRepositoryImpl : ITripRepository
{
    private static readonly ILog log =
        LogManager.GetLogger(System.Reflection.MethodBase.GetCurrentMethod().DeclaringType);

    private IValidator<Trip>? _tripValidator;

    private IDictionary<String, String> properties;

    public TripRepositoryImpl(IDictionary<String, String> properties, IValidator<Trip>? tripValidator)
    {
        log.Info("Creating TripDatabaseRepository");

        this.properties = properties;
        this._tripValidator = tripValidator;
    }
    
    private Trip GetTripFromResult(IDataReader dataReader)
    {
        int id = Convert.ToInt32(dataReader["id"]);
        string destination = dataReader["destination"].ToString();
        string transportCompany = dataReader["transportCompany"].ToString();
        string departureDate = dataReader["departureDate"].ToString().Trim();
        double price = Convert.ToDouble(dataReader["price"]);
        int numberOfAvailableSeats = Convert.ToInt32(dataReader["numberOfAvailableSeats"]);

        DateTime departure = DateTime.ParseExact(departureDate, "yyyy-MM-dd HH:mm", System.Globalization.CultureInfo.InvariantCulture);
        
        Trip? trip = new Trip(destination, transportCompany, departure, price, numberOfAvailableSeats);
        trip.Id = id;
        return trip;
    }

    public Trip? FindOne(int id)
    {
        log.Info("Finding Trip");

        IDbConnection connection = DBUtils.getConnection(properties);

        if (id == null)
        {
            throw new RepositoryException("Trip Id is null");
        }

        Trip? trip = null;
        using (var command = connection.CreateCommand())
        {
            command.CommandText = "select * from trip where id = @id";
            IDbDataParameter parameter = command.CreateParameter();
            parameter.ParameterName = "@id";
            parameter.Value = id;

            command.Parameters.Add(parameter);

            using (var dataReader = command.ExecuteReader())
            {
                if (dataReader.Read())
                {
                    trip = GetTripFromResult(dataReader);
                    log.InfoFormat("Exiting FindOne Trip with value: " + trip);
                }
            }
        }

        log.InfoFormat("Exiting FindOne Trip with value: " + trip);
        return trip;
    }

    public IEnumerable<Trip> FindAll()
    {
        log.Info("Finding All Trips");
        IDbConnection connection = DBUtils.getConnection(properties);
        IList<Trip> trips = new List<Trip>();

        using (var command = connection.CreateCommand())
        {
            command.CommandText = "select * from trip";

            using (var dataReader = command.ExecuteReader())
            {
                while (dataReader.Read())
                {
                    Trip trip = GetTripFromResult(dataReader);
                    trips.Add(trip);
                }
            }
        }

        log.InfoFormat("Exiting FindAll trips with value: " + trips);
        return trips;
    }
    
    public Trip? FindTrip(string destination, string transportCompany, DateTime departureDate, double price,
        int numberOfAvailableSeats)
    {
        log.Info("Finding Trip by all");
        IDbConnection connection = DBUtils.getConnection(properties);

        if (destination == "" || transportCompany == "" || departureDate == null || price == null || numberOfAvailableSeats == null)
        {
            throw new RepositoryException("Agency destination or transport company or departure dates or price or numberOfAvailableSeats are required");
        }
        
        Trip? trip = null;
        using (var command = connection.CreateCommand())
        {
            command.CommandText = "select * from trip where destination=@destionation and transportCompany=@transportCompany and departureDate=@departureDate and price=@price and numberOfAvailableSeats=@numberOfAvailableSeats";
            IDbDataParameter destinationParameter = command.CreateParameter();
            destinationParameter.ParameterName = "@destionation";
            destinationParameter.Value = destination;
            command.Parameters.Add(destinationParameter);
            
            IDbDataParameter transportParameter = command.CreateParameter();
            transportParameter.ParameterName = "@transportCompany";
            transportParameter.Value = transportCompany;
            command.Parameters.Add(transportParameter);
            
            IDbDataParameter departureDateParameter = command.CreateParameter();
            departureDateParameter.ParameterName = "@departureDate";
            departureDateParameter.Value = departureDate.ToString("yyyy-MM-dd HH:mm");;
            command.Parameters.Add(departureDateParameter);
            
            IDbDataParameter priceParameter = command.CreateParameter();
            priceParameter.ParameterName = "@price";
            priceParameter.Value = price;
            command.Parameters.Add(priceParameter);
            
            IDbDataParameter numberOfAvailableSeatsParameter = command.CreateParameter();
            numberOfAvailableSeatsParameter.ParameterName = "@numberOfAvailableSeats";
            numberOfAvailableSeatsParameter.Value = numberOfAvailableSeats;
            command.Parameters.Add(numberOfAvailableSeatsParameter);

            using (var dataReader = command.ExecuteReader())
            {
                if (dataReader.Read())
                {
                    trip = GetTripFromResult(dataReader);
                    log.InfoFormat("Exiting Find Trip with value: " + trip);
                    return trip;
                }
            }
        }
        log.InfoFormat("Exiting Find Trip with value: " + trip);
        return trip;
    }
    
     public Trip? Save(Trip trip)
    {
        log.Info("Saving Trip");
        IDbConnection connection = DBUtils.getConnection(properties);

        _tripValidator.Validate(trip);

        Trip? tripFound = FindTrip(trip.Destination, trip.TransportCompany, trip.DepartureDate, trip.Price, trip.NumberOfAvailableSeats);
        if (tripFound != null)
        {
            log.InfoFormat("Trip already exists with value: " + tripFound);
            throw new RepositoryException("Trip already exists");
        }

        using (var command = connection.CreateCommand())
        {
            command.CommandText =
                "insert into trip (destination, transportCompany, departureDate, price, numberOfAvailableSeats) values (@destination, @transportCompany, @departureDate, @price, @numberOfAvailableSeats)";
            IDbDataParameter destinationParameter = command.CreateParameter();
            destinationParameter.ParameterName = "@destination";
            destinationParameter.Value = trip.Destination;
            command.Parameters.Add(destinationParameter);
            
            IDbDataParameter transportParameter = command.CreateParameter();
            transportParameter.ParameterName = "@transportCompany";
            transportParameter.Value = trip.TransportCompany;
            command.Parameters.Add(transportParameter);
            
            IDbDataParameter departureDateParameter = command.CreateParameter();
            departureDateParameter.ParameterName = "@departureDate";
            departureDateParameter.Value = trip.DepartureDate.ToString("yyyy-MM-dd HH:mm");;
            command.Parameters.Add(departureDateParameter);
            
            IDbDataParameter priceParameter = command.CreateParameter();
            priceParameter.ParameterName = "@price";
            priceParameter.Value = trip.Price;
            command.Parameters.Add(priceParameter);
            
            IDbDataParameter numberOfAvailableSeatsParameter = command.CreateParameter();
            numberOfAvailableSeatsParameter.ParameterName = "@numberOfAvailableSeats";
            numberOfAvailableSeatsParameter.Value = trip.NumberOfAvailableSeats;
            command.Parameters.Add(numberOfAvailableSeatsParameter);

            command.ExecuteNonQuery();

            Trip? tripF = FindTrip(trip.Destination, trip.TransportCompany, trip.DepartureDate, trip.Price, trip.NumberOfAvailableSeats);
            trip.Id = tripF.Id;
            log.InfoFormat("Trip saved with value: " + trip);
        }
        return trip;
    }

    public Trip? Delete(int id)
    {
        log.Info("Deleting Trip");
        IDbConnection connection = DBUtils.getConnection(properties);

        if (id == null)
        {
            throw new RepositoryException("Trip Id is null");
        }
        
        Trip? tripFound = FindOne(id);
        if (tripFound == null)
        {
            log.InfoFormat("Trip is not found");
            throw new RepositoryException("Trip is not found");
        }

        using (var command = connection.CreateCommand())
        {
            command.CommandText = "delete from trip where id = @id";
            IDbDataParameter parameter = command.CreateParameter();
            parameter.ParameterName = "@id";
            parameter.Value = id;
            command.Parameters.Add(parameter);
        
            command.ExecuteNonQuery();

            log.InfoFormat("Trip deleted with value: " + tripFound);
        }

        return tripFound;
    }

    public Trip? Update(Trip trip)
    {
        log.Info("Updating Trip");
        IDbConnection connection = DBUtils.getConnection(properties);

        _tripValidator.Validate(trip);

        Trip? tripFound = FindOne(trip.Id);
        if (tripFound == null)
        { 
            log.InfoFormat("Trip is not found");
            throw new RepositoryException("Trip is not found");
        }

        using (var command = connection.CreateCommand())
        {
            command.CommandText = "update trip set destination=@destionation, transportCompany=@transportCompany, departureDate=@departureDate, price=@price, numberOfAvailableSeats=@numberOfAvailableSeats where id = @id";
            IDbDataParameter destinationParameter = command.CreateParameter();
            destinationParameter.ParameterName = "@destionation";
            destinationParameter.Value = trip.Destination;
            command.Parameters.Add(destinationParameter);
            
            IDbDataParameter transportParameter = command.CreateParameter();
            transportParameter.ParameterName = "@transportCompany";
            transportParameter.Value = trip.TransportCompany;
            command.Parameters.Add(transportParameter);
            
            IDbDataParameter departureDateParameter = command.CreateParameter();
            departureDateParameter.ParameterName = "@departureDate";
            departureDateParameter.Value = trip.DepartureDate.ToString("yyyy-MM-dd HH:mm");;
            command.Parameters.Add(departureDateParameter);
            
            IDbDataParameter priceParameter = command.CreateParameter();
            priceParameter.ParameterName = "@price";
            priceParameter.Value = trip.Price;
            command.Parameters.Add(priceParameter);
            
            IDbDataParameter numberOfAvailableSeatsParameter = command.CreateParameter();
            numberOfAvailableSeatsParameter.ParameterName = "@numberOfAvailableSeats";
            numberOfAvailableSeatsParameter.Value = trip.NumberOfAvailableSeats;
            command.Parameters.Add(numberOfAvailableSeatsParameter);

            IDbDataParameter parameter = command.CreateParameter();
            parameter.ParameterName = "@id";
            parameter.Value = trip.Id;
            command.Parameters.Add(parameter);
            
            command.ExecuteNonQuery();
            
            log.InfoFormat("Trip added with value: " + trip);
        }

        return trip;
    }

    public Trip? UpdateSeats(int id, int seatsReserved)
    {
        log.Info("Updating Seats");
        IDbConnection connection = DBUtils.getConnection(properties);
        Trip? tripFound = FindOne(id);
        
        if (tripFound == null)
        { 
            log.InfoFormat("Trip is not found");
            throw new RepositoryException("Trip is not found");
        }

        using (var command = connection.CreateCommand())
        {
            int numberOfAvailableSeats = tripFound.NumberOfAvailableSeats - seatsReserved;
            command.CommandText = "update trip set numberOfAvailableSeats=@numberOfAvailableSeats where id = @id";
            
            IDbDataParameter numberOfAvailableSeatsParameter = command.CreateParameter();
            numberOfAvailableSeatsParameter.ParameterName = "@numberOfAvailableSeats";
            numberOfAvailableSeatsParameter.Value = numberOfAvailableSeats;
            command.Parameters.Add(numberOfAvailableSeatsParameter);

            IDbDataParameter parameter = command.CreateParameter();
            parameter.ParameterName = "@id";
            parameter.Value = id;
            command.Parameters.Add(parameter);
            
            command.ExecuteNonQuery();
            
            Trip? tripF = FindOne(id);
            log.InfoFormat("Trip added with id: " + id);
            return tripF;
        }

        return null;
    }

    public IEnumerable<Trip> FindTripsByDestination(string destination)
    {
        log.Info("Finding Trips by destination");
        IDbConnection connection = DBUtils.getConnection(properties);
        IList<Trip> trips = new List<Trip>();

        using (var command = connection.CreateCommand())
        {
            command.CommandText = "select * from trip where destination=@destination";
            IDbDataParameter parameter = command.CreateParameter();
            parameter.ParameterName = "@destination";
            parameter.Value = destination;
            command.Parameters.Add(parameter);

            using (var dataReader = command.ExecuteReader())
            {
                while (dataReader.Read())
                {
                    Trip trip = GetTripFromResult(dataReader);
                    trips.Add(trip);
                }
            }
        }

        log.InfoFormat("Exiting FindAll trips with value: " + trips);
        return trips;
    }
}

File: Fourth Semester - Design and Programming Environments/Java/ProiectJava/java/org/example/repository/impl/TripRepositoryImpl.java
package org.example.repository.impl;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.example.exception.RepositoryException;
import org.example.modul.Trip;
import org.example.modul.validator.Validator;
import org.example.repository.interfaces.TripRepository;
import org.example.repository.utils.JdbcUtils;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.*;

public class TripRepositoryImpl implements TripRepository {
    protected JdbcUtils jdbcUtils;
    protected Validator<Trip> validator;
    private static final Logger logger = LogManager.getLogger();

    public TripRepositoryImpl(Properties properties, Validator<Trip> validator) {
        logger.info("Initializing TripDatabaseRepository with properties: {}", properties);

        this.jdbcUtils = new JdbcUtils(properties);
        this.validator = validator;
    }

    private Trip getTripFromSet(ResultSet result) throws SQLException {
        Integer id = result.getInt("id");
        String destination = result.getString("destination");
        String transportCompany = result.getString("transportCompany");
        String departure = result.getString("departureDate");
        Double price = result.getDouble("price");
        Integer numberOfAvailableSeats = result.getInt("numberOfAvailableSeats");

        DateTimeFormatter formatter = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm");
        LocalDateTime departureTime = LocalDateTime.parse(departure, formatter);

        Trip trip = new Trip(destination, transportCompany, departureTime, price, numberOfAvailableSeats);
        trip.setId(id);
        return trip;
    }

    @Override
    public Iterable<Trip> findAll() {
        logger.trace("Finding all trips");

        Connection connection = jdbcUtils.getConnection();

        List<Trip> trips = new ArrayList<>();
        try(PreparedStatement preparedStatement = connection.prepareStatement("select * from trip")){
            try(ResultSet result = preparedStatement.executeQuery()){
                while (result.next()){
                    Trip trip = getTripFromSet(result);
                    trips.add(trip);
                }
            }
        } catch (Exception e) {
            logger.error(e);
            System.out.println("Error getting connection "+e);
        }

        logger.traceExit("Found trips: {}", trips);
        return trips;
    }

    @Override
    public Optional<Trip> findOne(Integer id) {
        logger.trace("Finding trip with id {}", id);

        Connection connection = jdbcUtils.getConnection();

        Trip trip = null;
        try(PreparedStatement preparedStatement = connection.prepareStatement("select * from trip where id = ?")){
            preparedStatement.setInt(1, id);
            try(ResultSet result = preparedStatement.executeQuery()){
                if (result.next()){
                    trip = getTripFromSet(result);
                    logger.traceExit("Found trip: {}", trip);
                }
                else {
                    logger.warn("No trip found with id {}", id);
                }
            }
        } catch (SQLException e) {
            logger.error(e);
        }

        logger.traceExit();
        return Optional.ofNullable(trip);
    }

    @Override
    public Optional<Trip> save(Trip entity) {
        logger.trace("Saving trip {}", entity);

        validator.validate(entity);

        Connection connection = jdbcUtils.getConnection();

        try(PreparedStatement preparedStatement = connection.prepareStatement("insert into trip (destination, transportCompany, departureDate, price, numberOfAvailableSeats) values (?, ?, ?, ?, ?)")){
            preparedStatement.setString(1, entity.getDestination());
            preparedStatement.setString(2, entity.getTransportCompany());
            preparedStatement.setString(3, entity.getDepartureDate().format(DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm")));
            preparedStatement.setDouble(4, entity.getPrice());
            preparedStatement.setInt(5, entity.getNumberOfAvailableSeats());
            int result = preparedStatement.executeUpdate();
            if(result == 0) {
                logger.warn("Failed to save trip {}", entity);
            }
            else {
                Optional<Trip> trip = findTrip(entity.getDestination(), entity.getTransportCompany(),
                        entity.getDepartureDate(), entity.getPrice(), entity.getNumberOfAvailableSeats());
                entity.setId(trip.get().getId());
                logger.trace("Successfully saved trip {}", entity);
            }
        } catch (SQLException exception) {
            logger.error(exception);
            throw new RepositoryException("Error saving trip in database", exception);
        }

        logger.traceExit();
        return Optional.of(entity);
    }

    @Override
    public Optional<Trip> delete(Integer id) {
        logger.trace("Deleting trip with id {}", id);

        Connection connection = jdbcUtils.getConnection();

        Optional<Trip> trip = findOne(id);
        if(trip.isEmpty()) {
            throw new RepositoryException("This trip doesn't exists");
        }

        try(PreparedStatement preparedStatement = connection.prepareStatement("delete from trip where id = ?")){
            preparedStatement.setInt(1, id);

            int result = preparedStatement.executeUpdate();
            if(result == 0) {
                logger.warn("Failed to delete trip {}", trip);
            }
            else {
                logger.trace("Successfully deleted trip {}", trip);
            }
        } catch (SQLException e) {
            logger.error(e);
            throw new RepositoryException("Error deleting trip in database", e);
        }

        logger.traceExit();
        return trip;
    }

    @Override
    public Optional<Trip> update(Trip entity) {
        logger.trace("Updating trip {}", entity);

        validator.validate(entity);

        Connection connection = jdbcUtils.getConnection();

        try(PreparedStatement preparedStatement = connection.prepareStatement("update trip set destination = ?, " +
                "transportCompany = ?, departureDate = ?, price = ?, numberOfAvailableSeats = ? where id = ?")){
            preparedStatement.setString(1, entity.getDestination());
            preparedStatement.setString(2, entity.getTransportCompany());
            preparedStatement.setString(3, entity.getDepartureDate().format(DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm")));
            preparedStatement.setDouble(4, entity.getPrice());
            preparedStatement.setInt(5, entity.getNumberOfAvailableSeats());
            preparedStatement.setInt(6, entity.getId());

            int result = preparedStatement.executeUpdate();
            if(result == 0) {
                logger.warn("Failed to update trip {}", entity);
            }
            else {
                logger.trace("Successfully updated trip {}", entity);
            }
        } catch (SQLException e) {
            logger.error(e);
            throw new RepositoryException("Error updating trip in database", e);
        }

        logger.traceExit();
        return Optional.of(entity);
    }

    @Override
    public Optional<Trip> findTrip(String destination, String transportCompany, LocalDateTime departureDate, Double price, Integer numberOfAvailableSeats) {
        logger.trace("Finding trip with destination {}, transportCompany {}, departureDate {}, price {}, numberOfAvailableSeats {}", destination, transportCompany, departureDate, price, numberOfAvailableSeats);

        Connection connection = jdbcUtils.getConnection();

        Trip trip = null;
        try(PreparedStatement preparedStatement = connection.prepareStatement("select * from trip where destination = ? " +
                "and transportCompany = ? and departureDate = ? and price = ? and numberOfAvailableSeats = ?")){
            preparedStatement.setString(1, destination);
            preparedStatement.setString(2, transportCompany);
            preparedStatement.setString(3, departureDate.format(DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm")));
            preparedStatement.setDouble(4, price);
            preparedStatement.setInt(5, numberOfAvailableSeats);
            try(ResultSet result = preparedStatement.executeQuery()){
                if (result.next()){
                    trip = getTripFromSet(result);
                    logger.trace("Found trip: {}", trip);
                }
                else {
                    logger.warn("No trip found with destination {}, transportCompany {}, departureDate {}, price {}, numberOfAvailableSeats {}", destination, transportCompany, departureDate, price, numberOfAvailableSeats);
                }
            }
        } catch (SQLException e) {
            logger.error(e);
        }

        logger.traceExit();
        return Optional.ofNullable(trip);
    }

    @Override
    public Iterable<Trip> findTripsByDestination(String destination) {
        logger.trace("Finding trips with destination {}", destination);

        Connection connection = jdbcUtils.getConnection();

        List<Trip> trips = new ArrayList<>();
        try(PreparedStatement preparedStatement = connection.prepareStatement("select * from trip where destination = ?")){
            preparedStatement.setString(1, destination);
            try(ResultSet result = preparedStatement.executeQuery()){
                while (result.next()){
                    Trip trip = getTripFromSet(result);
                    trips.add(trip);
                }
            }
        } catch (SQLException e) {
            logger.error(e);
        }

        logger.traceExit("Found trips: {}", trips);
        return trips;
    }

    @Override
    public Optional<Trip> updateSeats(Integer id, Integer numberOfReservedSeats) {
        logger.trace("Updating trip with id {}", id);

        Connection connection = jdbcUtils.getConnection();

        Optional<Trip> trip = findOne(id);
        if(trip.isEmpty()) {
            throw new RepositoryException("This trip doesn't exists");
        }

        try(PreparedStatement preparedStatement = connection.prepareStatement("update trip set numberOfAvailableSeats = ? where id = ?")){
            Integer numberOfAvailableSeats = trip.get().getNumberOfAvailableSeats() - numberOfReservedSeats;
            preparedStatement.setInt(1, numberOfAvailableSeats);
            preparedStatement.setInt(2, id);

            int result = preparedStatement.executeUpdate();
            if(result == 0) {
                logger.warn("Failed to update trip {}", trip);
            }
            else {
                logger.trace("Successfully updated trip {}", trip);
            }
        } catch (SQLException e) {
            logger.error(e);
            throw new RepositoryException("Error updating trip in database", e);
        }

        logger.traceExit();
        return trip;
    }
}
